File: tests/fixtures/vuln_app.js
/**
 * Deliberately vulnerable JavaScript application for testing source analysis.
 *
 * DO NOT deploy this code. It contains intentional security vulnerabilities
 * for testing skwaq's multi-language source code analysis.
 */

const express = require('express');
const child_process = require('child_process');
const fs = require('fs');

const app = express();

// Vulnerability 1: Command injection via eval
function handleSearch(req, res) {
    const query = req.params.query;
    const result = eval(query);
    res.send(result);
}

// Vulnerability 2: Command injection via child_process.exec
function runCommand(req, res) {
    const cmd = req.body.command;
    child_process.exec(cmd, function(err, stdout) {
        res.send(stdout);
    });
}

// Vulnerability 3: XSS via innerHTML
function renderProfile(req, res) {
    const name = req.query.name;
    const html = '<div>' + name + '</div>';
    document.innerHTML = html;
    document.write(html);
}

// Vulnerability 4: Dynamic code execution via new Function
function compute(req, res) {
    const expression = req.body.expr;
    const fn = new Function('return ' + expression);
    res.send(fn());
}

// Vulnerability 5: Prototype pollution
function merge(target, source) {
    for (const key in source) {
        if (key === '__proto__') {
            target[key] = source[key];
        }
        if (typeof source[key] === 'object') {
            target[key] = merge(target[key] || {}, source[key]);
        } else {
            target[key] = source[key];
        }
    }
    return target;
}

// Vulnerability 6: eval-equivalent via setTimeout string
function delayedExec(req, res) {
    const code = req.params.code;
    setTimeout(code, 1000);
    setInterval(code, 5000);
}

// Vulnerability 7: SQL injection (if using a SQL driver)
function getUser(req, res) {
    const id = req.params.id;
    const env_key = process.env.DB_KEY;
    db.query("SELECT * FROM users WHERE id = " + id);
}

// Vulnerability 8: File write with user input
function saveFile(req, res) {
    const content = req.body.content;
    const path = req.body.path;
    fs.writeFile(path, content, function(err) {
        res.send("saved");
    });
}

app.get('/search/:query', handleSearch);
app.post('/run', runCommand);
app.get('/profile', renderProfile);
app.post('/compute', compute);
app.get('/delay/:code', delayedExec);
app.get('/user/:id', getUser);
app.post('/save', saveFile);

app.listen(3000);

File: tests/fixtures/vuln_app.py
"""Deliberately vulnerable Python application for testing source analysis.

DO NOT deploy this code. It contains intentional security vulnerabilities
for testing skwaq's multi-language source code analysis.
"""

import os
import pickle
import subprocess
import sqlite3
import yaml


def handle_request(request):
    """Handle an HTTP request with multiple vulnerabilities."""
    # Source: HTTP input
    name = request.args.get("name")
    query = request.args.get("query")
    data = request.data

    # Vulnerability 1: SQL Injection via string concatenation
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users WHERE name = '" + name + "'")

    # Vulnerability 2: Command Injection via os.system
    os.system("echo " + name)

    # Vulnerability 3: Code Injection via eval
    result = eval(query)

    # Vulnerability 4: Code Injection via exec
    exec(data)

    return result


def load_user_data(file_path):
    """Load user data from a file with deserialization vulnerability."""
    # Source: File read
    with open(file_path, "rb") as f:
        raw = f.read()

    # Vulnerability 5: Unsafe deserialization via pickle
    user = pickle.loads(raw)

    # Vulnerability 6: Unsafe YAML loading
    config = yaml.load(raw)

    return user, config


def run_command(user_input):
    """Run a command based on user input."""
    # Source: User input
    cmd = input("Enter command: ")

    # Vulnerability 7: Command injection via subprocess
    subprocess.call(cmd, shell=True)
    subprocess.Popen(cmd, shell=True)

    # Vulnerability 8: Dynamic import
    module = __import__(user_input)

    return module


def get_env_config():
    """Read configuration from environment."""
    # Source: Environment variables
    secret = os.environ.get("SECRET_KEY")
    db_url = os.environ.get("DATABASE_URL")

    # Vulnerability 9: SQL injection via environment variable
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM config WHERE key = '" + db_url + "'")

    return secret


def write_output(data, path):
    """Write data to a file."""
    # Sink: File write
    with open(path, "w") as f:
        f.write(data)
